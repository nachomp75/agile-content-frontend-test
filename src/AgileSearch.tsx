import { AppRouter } from './router/AppRouter'

function AgileSearch() {
  return (
    <AppRouter />
  )
}

export default AgileSearch
