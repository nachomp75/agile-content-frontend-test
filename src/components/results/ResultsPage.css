.results-loading,
.results-page {
  box-sizing: border-box;
  flex: 1 1 auto;
  height: 100%;
  padding: 30px 50px;
}

.results-page {
  display: grid;
  grid-template-columns: 2fr 1fr;
}

@media (max-width: 992px) {
  .results-page {
    display: block;
  }
}

@media (max-width: 576px) {
  .results-page {
    padding: 30px;
  }
}
