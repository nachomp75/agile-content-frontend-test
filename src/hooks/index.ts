export * from './useClickOutside'
export * from './useResults'
export * from './useSearch'
