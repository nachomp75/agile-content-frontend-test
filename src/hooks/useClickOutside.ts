import { useEffect, useRef } from 'react'

export const useClickOutside = (callback: () => void) => {
  const ref = useRef<HTMLElement>(null)

  useEffect(() => {
    const handleClick = ({ target }: MouseEvent) => {
      if (!ref.current?.contains(target as Node)) {
        callback()
      }
    }

    document.addEventListener('click', handleClick, true)

    return () => {
      document.removeEventListener('click', handleClick, true)
    }
  }, [ref])

  return ref
}
