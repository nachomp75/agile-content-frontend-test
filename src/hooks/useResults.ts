import { useEffect, useState } from 'react'
import { useLocation } from 'react-router-dom'

import { getResults } from '../services'
import { type Result } from '../interfaces'

export const useResults = () => {
  const [searchTerm, setSearchTerm] = useState('')
  const [results, setResults] = useState<Result[]>([])
  const [loading, setLoading] = useState(false)
  const [selectedResult, setSelectedResult] = useState<Result>()
  const { search } = useLocation()

  useEffect(() => {
    const term = search.split('=')[1]
    setSearchTerm(term)

    if (term) {
      setLoading(true)

      getResults(term)
        .then((results) => {
          setResults(results)
          setLoading(false)
        })
        .catch(() => {
          setLoading(false)
        })
    } else {
      setResults([])
    }
  }, [search])

  return {
    loading,
    results,
    searchTerm,
    selectedResult,
    setSelectedResult
  }
}
