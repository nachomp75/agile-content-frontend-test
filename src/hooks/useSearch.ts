import {
  useEffect,
  useState,
  type ChangeEvent,
  type FormEvent,
  type KeyboardEvent
} from 'react'
import { useLocation, useNavigate } from 'react-router-dom'

export const useSearch = () => {
  const navigate = useNavigate()
  const { search } = useLocation()
  const [searchValue, setSearchValue] = useState('')

  const handleSubmit = (e?: FormEvent) => {
    e?.preventDefault()

    const searchText = searchValue.trim()
    navigate(`/results${searchText && `?q=${searchText}`}`)
  }

  const handleInputChange = ({ target }: ChangeEvent<HTMLInputElement>) => {
    setSearchValue(target.value)
  }

  const handleSubmitOnEnter = ({ key }: KeyboardEvent) => {
    if (key === 'Enter') handleSubmit()
  }

  useEffect(() => {
    const term = search.split('=')[1]
    setSearchValue(term || '')
  }, [search])

  return {
    searchValue,
    setSearchValue,
    handleInputChange,
    handleSubmitOnEnter,
    handleSubmit
  }
}
