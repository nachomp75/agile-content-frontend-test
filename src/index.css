:root {
  --background-primary: white;
  --background-secondary: lavender;
  --background-tertiary: ghostwhite;
  --color-font-primary: #242424;
  --color-font-secondary: #bbbbbb;
  --shadow-default: rgba(149, 157, 165, 0.2) 0px 8px 24px;

  font-family: Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
  font-weight: 400;
  color: var(--color-font-primary);
}

body {
  margin: 0;
}

#root {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  min-height: 100vh;
}
