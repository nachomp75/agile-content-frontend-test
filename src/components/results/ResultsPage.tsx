import { ResultsList } from './ResultsList'
import { ResultCard } from './ResultCard'
import { useResults } from '../../hooks'
import { ResultSkeleton } from '../shared'

import './ResultsPage.css'

export const ResultsPage = () => {
  const { loading, results, searchTerm, selectedResult, setSelectedResult } =
    useResults()

  if (loading) {
    return (
      <div className='results-loading' data-testid='results-skeleton'>
        {[...new Array(4)].map((_item, index) => (
          <ResultSkeleton key={index} />
        ))}
      </div>
    )
  }

  return (
    <div className='results-page'>
      <div className='results'>
        <ResultsList
          results={results}
          searchTerm={searchTerm}
          setSelectedResult={setSelectedResult}
        />
      </div>

      {selectedResult && (
        <ResultCard
          result={selectedResult}
          setSelectedResult={setSelectedResult}
        />
      )}
    </div>
  )
}

export default ResultsPage
