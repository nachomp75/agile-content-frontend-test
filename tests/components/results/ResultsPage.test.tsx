import { render, screen } from '@testing-library/react'

import ResultsPage from '../../../src/components/results/ResultsPage'
import { useResults } from '../../../src/hooks'
import { type Result } from '../../../src/interfaces'

jest.mock('../../../src/hooks/useResults')

const useResultsMock = useResults as jest.MockedFunction<typeof useResults>

describe('<ResultsPage />', () => {
  const resultMock: Result = {
    type: '',
    id: 1,
    url: '',
    title: 'Bear',
    description: '',
    image: ''
  }

  test('should match snapshot', () => {
    useResultsMock.mockReturnValue({
      loading: true,
      results: [],
      searchTerm: '',
      selectedResult: undefined,
      setSelectedResult: jest.fn()
    })
    const { container } = render(<ResultsPage />)
    expect(container).toMatchSnapshot()
  })

  test('should render results skeleton while loading', () => {
    useResultsMock.mockReturnValue({
      loading: true,
      results: [],
      searchTerm: '',
      selectedResult: undefined,
      setSelectedResult: jest.fn()
    })
    render(<ResultsPage />)

    const resultsSkeleton = screen.getByTestId('results-skeleton')
    expect(resultsSkeleton).not.toBeNull()
  })

  test('should render <ResultsList /> when not loading', () => {
    useResultsMock.mockReturnValue({
      loading: false,
      results: [resultMock],
      searchTerm: '',
      selectedResult: undefined,
      setSelectedResult: jest.fn()
    })
    render(<ResultsPage />)

    const resultsList = screen.getByText(resultMock.title)
    expect(resultsList).not.toBeNull()
  })

  test('should render <ResultCard /> when result is selected', () => {
    useResultsMock.mockReturnValue({
      loading: false,
      results: [resultMock],
      searchTerm: '',
      selectedResult: resultMock,
      setSelectedResult: jest.fn()
    })
    render(<ResultsPage />)

    const resultsList = screen.getAllByRole('figure')
    expect(resultsList).not.toBeNull()
  })
})
