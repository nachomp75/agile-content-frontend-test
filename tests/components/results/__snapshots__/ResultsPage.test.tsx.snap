// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultsPage /> should match snapshot 1`] = `
<div>
  <div
    class="results-loading"
    data-testid="results-skeleton"
  >
    <div
      class="skeleton-loader"
    >
      <div
        class="section1"
      />
      <div
        class="section2"
      />
      <div
        class="section3"
      />
    </div>
    <div
      class="skeleton-loader"
    >
      <div
        class="section1"
      />
      <div
        class="section2"
      />
      <div
        class="section3"
      />
    </div>
    <div
      class="skeleton-loader"
    >
      <div
        class="section1"
      />
      <div
        class="section2"
      />
      <div
        class="section3"
      />
    </div>
    <div
      class="skeleton-loader"
    >
      <div
        class="section1"
      />
      <div
        class="section2"
      />
      <div
        class="section3"
      />
    </div>
  </div>
</div>
`;
